bqskit/runtime/__init__.py

```diff
@@ -0,0 +1,183 @@
+ip address and optionally ports. You can see the `-h` option of each
+command or the :func:`~bqskit.runtime.detached.start_server` and
+:func:`~bqskit.runtime.manager.start_manager` entry points for
+more information.
+
+For more information on how to manage jobs submitted to a runtime server,
+see the :obj:`~bqskit.compiler` documentation.
+
+.. autosummary::
+    :toctree: autogen
+    :recursive:
+
+    ~bqskit.runtime.attached.AttachedServer
+    ~bqskit.runtime.detached.DetachedServer
+    ~bqskit.runtime.manager.Manager
+    ~bqskit.runtime.detached.start_server
+    ~bqskit.runtime.manager.start_manager
+
+Parallelize Pass Computation
+############################
+
+When developing a BQSKit compiler pass, you can hook into the active
+runtime server through the :func:`get_runtime` function. This returns a
+:class:`RuntimeHandle`, which you can use to submit, map, wait on, and
+cancel tasks in the execution environment.
+
+For more information on how to design a custom pass, see this (TODO, sorry,
+you can look at the source code of existing
+`passes <https://github.com/BQSKit/bqskit/blob/master/bqskit/passes>`_
+for a good example for the time being).
+
+.. autosummary::
+    :toctree: autogen
+    :recursive:
+
+    ~bqskit.runtime.get_runtime
+    ~bqskit.runtime.RuntimeHandle
+    ~bqskit.runtime.future.RuntimeFuture
+
+Standard Logging Interface
+##########################
+
+This system uses the standard python logging module to provide a familiar
+logging interface. On client processes -- the ones where you create the
+Compiler object -- you can configure python loggers like normal before
+submitting tasks and have the entire system honor that configuration.
+"""
+from __future__ import annotations
+
+import faulthandler
+import os
+from typing import Any
+from typing import Callable
+from typing import Protocol
+from typing import TYPE_CHECKING
+
+# Enable low-level fault handling: system crashes print a minimal trace.
+faulthandler.enable()
+
+
+# Disable multi-threading in BLAS libraries.
+os.environ['OMP_NUM_THREADS'] = '1'
+os.environ['OPENBLAS_NUM_THREADS'] = '1'
+os.environ['MKL_NUM_THREADS'] = '1'
+os.environ['NUMEXPR_NUM_THREADS'] = '1'
+os.environ['VECLIB_MAXIMUM_THREADS'] = '1'
+os.environ['RUST_BACKTRACE'] = '1'
+
+
+if TYPE_CHECKING:
+    from bqskit.runtime.future import RuntimeFuture
+
+
+class RuntimeHandle(Protocol):
+    """
+    A structural type capturing BQSKit Runtime capabilities.
+
+    A compiler pass designer can effortlessly parallelize certain aspects
+    of compilation using the :func:`submit` and :func:`map` methods.
+    These both will return :class:`~bqskit.runtime.future.RuntimeFuture`
+    objects, which can be awaited on using the standard python `await`
+    keyword. This can also be used to cancel tasks and wait on specific
+    events.
+
+    This should never be constructed directly and only accessed through
+    the :func:`get_runtime` function.
+    """
+
+    def submit(
```

Comment:
The `RuntimeHandle` is a structural type definition since it subclasses `Protocol`. You can read more [here](https://mypy.readthedocs.io/en/stable/protocols.html). I added a note to the documentation to clarify this.